txnsync/txngroups.go

```diff
@@ -60,6 +454,953 @@ func decodeTransactionGroups(bytes []byte) (txnGroups []transactions.SignedTxGro
+
+	return txnGroups, nil
+}
+
+func addGroupHashes(txnGroups []transactions.SignedTxGroup, txnCount int, b bitmask) {
+	index := 0
+	for _, txns := range txnGroups {
+		var txGroup transactions.TxGroup
+		txGroup.TxGroupHashes = make([]crypto.Digest, len(txns.Transactions))
+		for i, tx := range txns.Transactions {
+			txGroup.TxGroupHashes[i] = crypto.HashObj(tx.Txn)
+		}
+		groupHash := crypto.HashObj(txGroup)
+		b.expandBitmask(txnCount)
+		for i := range txns.Transactions {
+			if exists := b.EntryExists(index); exists {
+				txns.Transactions[i].Txn.Group = groupHash
+			}
+			index++
+		}
+	}
+}
+
+func setupDeconstructSignedTransactions(stub *txGroupsEncodingStub) {
+	bitmaskLen := bytesNeededBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskAuthAddr = make(bitmask, bitmaskLen)
+	stub.AuthAddr = make([]byte, 0, int(stub.TotalTransactionsCount)*addressSize)
+	stub.BitmaskLsig = make(bitmask, bitmaskLen)
+	stub.Lsig = make([]transactions.LogicSig, 0, int(stub.TotalTransactionsCount))
+	stub.BitmaskMsig = make(bitmask, bitmaskLen)
+	stub.Msig = make([]crypto.MultisigSig, 0, int(stub.TotalTransactionsCount))
+	stub.BitmaskSig = make(bitmask, bitmaskLen)
+	stub.Sig = make([]byte, 0, int(stub.TotalTransactionsCount)*signatureSize)
+	setupDeconstructTransactions(stub)
+}
+
+// deconstructs SignedTxn's into lists of fields and bitmasks
+func deconstructSignedTransactions(stub *txGroupsEncodingStub, i int, txn transactions.SignedTxn) {
+
+	if !txn.Sig.MsgIsZero() {
+		stub.BitmaskSig.SetBit(i)
+		stub.Sig = append(stub.Sig, txn.Sig[:]...)
+	}
+	if !txn.Msig.MsgIsZero() {
+		stub.BitmaskMsig.SetBit(i)
+		stub.Msig = append(stub.Msig, txn.Msig)
+	}
+	if !txn.Lsig.MsgIsZero() {
+		stub.BitmaskLsig.SetBit(i)
+		stub.Lsig = append(stub.Lsig, txn.Lsig)
+	}
+	if !txn.AuthAddr.MsgIsZero() {
+		stub.BitmaskAuthAddr.SetBit(i)
+		stub.AuthAddr = append(stub.AuthAddr, txn.AuthAddr[:]...)
+	}
+	deconstructTransactions(stub, i, txn)
+}
+
+func finishDeconstructSignedTransactions(stub *txGroupsEncodingStub) {
+	stub.BitmaskAuthAddr.trimBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskLsig.trimBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskMsig.trimBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskSig.trimBitmask(int(stub.TotalTransactionsCount))
+	finishDeconstructTransactions(stub)
+}
+
+func setupDeconstructTransactions(stub *txGroupsEncodingStub) {
+	bitmaskLen := bytesNeededBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskTxType = make(bitmask, bitmaskLen)
+	stub.TxType = make([]byte, 0, int(stub.TotalTransactionsCount))
+	setupDeconstructTxnHeader(stub)
+	setupDeconstructKeyregTxnFields(stub)
+	setupDeconstructPaymentTxnFields(stub)
+	setupDeconstructAssetConfigTxnFields(stub)
+	setupDeconstructAssetTransferTxnFields(stub)
+	setupDeconstructAssetFreezeTxnFields(stub)
+	setupDeconstructApplicationCallTxnFields(stub)
+	setupDeconstructCompactCertTxnFields(stub)
```

Comment:
why not to switch on the type in order to make the right call ? It makes it cleaner, and reduce the extra conditional within each of these methods.